hal/pipeline: fix error codes

msune opened the pull request:
Please note that this change **breaks** 0.6 API; the problem is the bug is actually (partially) the HAL API itself for fm/gm. Therefore issuing against 0.6.

@vicalro review? To try it use the HEAD of this xDPd branch:

https://github.com/msune/xdpd/tree/fix_endpoint_return_codes

---

This commit addresses the problems related to the population
of errors from the HAL driver (originated, or not, from
rofl-pipeline) to the CMM.

It is important to note that HAL return code types are
intentionally not defined as typedefs of pipeline codes:

  typedef of1x_fm_result_t hal_fm_result_t;

because:
- HAL != pipeline, although id the driver uses the pipeline
  for flowmod/groupmod state cache, there might be a 1:1 mapping.
  Adding different types clearly separates them, to prevent
  confusion.
- HAL may have additional (non-pipeline related) return codes.

Defailed changelog:
- Defines missing TABLE_FULL/VALIDATION return codes in the
  pipeline.
- Defines missing hal_fm_result_t and hal_gm_result_t.
- Moves the return codes of flowmod operations to hal_fm_result_t
- Provides automatic conversion routines for pipeline->HAL return
  codes.
- Renamed ROFL_OF1X_GM_OK to ROFL_OF1X_GM_SUCCESS for consistency

Commit:
hal/pipeline: fix error codes

This commit addresses the problems related to the population
of errors from the HAL driver (originated, or not, from
rofl-pipeline) to the CMM.

It is important to note that HAL return code types are
intentionally not defined as typedefs of pipeline codes:

  typedef of1x_fm_result_t hal_fm_result_t;

because:

* HAL != pipeline, although id the driver uses the pipeline
  for flowmod/groupmod state cache, there might be a 1:1 mapping.
  Adding different types clearly separates them, to prevent
  confusion.
* HAL may have additional (non-pipeline related) return codes.

Defailed changelog:

- Defines missing TABLE_FULL/VALIDATION return codes in the
  pipeline.
- Defines missing hal_fm_result_t and hal_gm_result_t.
- Moves the return codes of flowmod operations to hal_fm_result_t
- Provides automatic conversion routines for pipeline->HAL return
  codes.
- Renamed ROFL_OF1X_GM_OK to ROFL_OF1X_GM_SUCCESS for consistency

Commit:
HAL: Adding missing INVALID_TABLE_ID return code

Some datapaths may have restrictions in the table ID where they
can accept certain flowmods. This patch adds this return code.

Commit:
hal/pipeline: fix modify/delete return codes

This patch fixes inconsistencies in the HAL and pipeline code for
modify and remove flowmod operations. For these operations, the pipeline
can and _should_ set the appropriate return codes too, the same way that
the add operation does.

msune commented:
Fixed the return codes for the pipeline and HAL.

To test use xdpd commit:

https://github.com/msune/xdpd/commit/4262de0d2d118c27f8ef9db27aac4d591e002dec

vicalro commented:
Checked the new error codes all working.
Found though that instruction type `OF1X_IT_GOTO_TABLE` its not being checked.
check https://github.com/bisdn/rofl-datapath/blob/stable/src/rofl/datapath/pipeline/openflow/openflow1x/pipeline/of1x_instruction.c#L256

Commit:
pipeline: validate table_id for GOTO_TABLE inst

This patch implements the validation of the table id in GOTO_TABLE
instructions, in order to make sure it is within the right range of
table ids.

vicalro commented:
Good, thanks!
ACK
